modules/modelSetup/mixin/ModelSetupDiffusionLossMixin.py

```diff
@@ -57,23 +64,58 @@ def _diffusion_loss(
             # TODO: don't disable masked loss functions when has_conditioning_image_input is true.
             #  This breaks if only the VAE is trained, but was loaded from an inpainting checkpoint
             if args.masked_training and not args.model_type.has_conditioning_image_input():
-                losses = loss_util.masked_loss(
-                    F.mse_loss,
-                    data['predicted'],
-                    data['target'],
-                    batch['latent_mask'],
-                    args.unmasked_weight,
-                    args.normalize_masked_area_loss
-                ).mean([1, 2, 3])
+            
+                #MSE/L2 Loss
+                if mse_strength != 0:
+                    mse_losses = loss_util.masked_loss(
+                        F.mse_loss,
+                        data['predicted'],
+                        data['target'],
+                        batch['latent_mask'],
+                        args.unmasked_weight,
+                        args.normalize_masked_area_loss
+                    ).mean([1, 2, 3])
+                    
+                #MAE/L1 Loss
+                if mae_strength != 0:
+                    mae_losses = loss_util.masked_loss(
+                        F.l1_loss,
+                        data['predicted'],
+                        data['target'],
+                        batch['latent_mask'],
+                        args.unmasked_weight,
+                        args.normalize_masked_area_loss
+                    ).mean([1, 2, 3])
             else:
-                losses = F.mse_loss(
-                    data['predicted'],
-                    data['target'],
-                    reduction='none'
-                ).mean([1, 2, 3])
+                
+                #MSE/L2 Loss
+                if mse_strength != 0:
+                    mse_losses = F.mse_loss(
+                        data['predicted'],
+                        data['target'],
+                        reduction='none'
+                    ).mean([1, 2, 3])
+
+                #MAE/L1 Loss
+                if mae_strength != 0:
+                    mae_losses = F.l1_loss(
+                        data['predicted'],
+                        data['target'],
+                        reduction='none'
+                    ).mean([1, 2, 3])
 
                 if args.masked_training and args.normalize_masked_area_loss:
                     clamped_mask = torch.clamp(batch['latent_mask'], args.unmasked_weight, 1)
-                    losses = losses / clamped_mask.mean(dim=(1, 2, 3))
+                    mse_losses = mse_losses / clamped_mask.mean(dim=(1, 2, 3))
+                    mae_losses = mae_losses / clamped_mask.mean(dim=(1, 2, 3))
```

Comment:
Does it make sense to only compute the mean once, and cache it?

Reply:
Good point. I changed it to a single calculation. It might be even better to cache the value. But that will require a change to the dataloader.